File: static/CWE-601/javascript/patched.js
// Remediation for CWE-601: URL Redirection to Untrusted Site ('Open Redirect')
// Fix: Only redirect to paths/hosts present in a server-side allowlist, rejecting any other target.

const express = require('express');

const app = express();

const ALLOWED_REDIRECTS = new Set(['/home', '/dashboard', '/login']);

app.get('/redirect', (req, res) => {
  const url = req.query.url;
  if (!ALLOWED_REDIRECTS.has(url)) {
    return res.status(400).send('invalid redirect target');
  }
  res.redirect(url);
});

app.listen(3000);

File: static/CWE-601/csharp/patched.cs
// Remediation for CWE-601: URL Redirection to Untrusted Site ('Open Redirect')
// Fix: Only allow local relative redirect targets, falling back to a safe default otherwise.

using System;
using System.Web;

namespace GuineaPig.Sast
{
    class Patched_601
    {
        public void Redirect(HttpRequest Request, HttpResponse Response)
        {
            string url = Request.QueryString["url"];
            string target = "/";

            if (!string.IsNullOrEmpty(url)
                && url.StartsWith("/", StringComparison.Ordinal)
                && !url.StartsWith("//", StringComparison.Ordinal)
                && !url.StartsWith("/\\", StringComparison.Ordinal))
            {
                target = url;
            }

            Response.Redirect(target);
        }
    }
}
